### ISSUE TITLE
Architecture to add libp2p tests

### ISSUE THREAD

**Issue by Elexy** [ipfs/benchmarks#198]

**Comment by vasco-santos** [comment]
Any update on this @Elexy @mcollina ?

**Comment by Elexy** [comment]
Working on this today.

**Comment by Elexy** [comment]
@vasco-santos I've added a branch with a design diagram.
https://github.com/ipfs/benchmarks/blob/add_libp2p_arch/runner/Libp2p%20Runner.png
The key differentiator would be the provisioning.
This takes care of setting up the right libs at the right version on the `Minion`.

**Comment by Elexy** [comment]
@vasco-santos I've made a [design diagram](https://github.com/ipfs/benchmarks/blob/add_libp2p_arch/runner/Libp2p%20Runner.png)

The main difference is the `Provisioning`. This is where we make sure the `minion` has the right libs at the right version before we run the benchmarks.

**Comment by vasco-santos** [comment]
Thanks for the proposal @Elexy ! 

So, in this approach we need to use this repo for all the tests and mark each test as being part of a project?

**Comment by Elexy** [comment]
Hi @vasco-santos
The way we get the library in the test bench is quite specific to this setup.
https://github.com/ipfs/benchmarks/blob/master/tests/getIpfs.sh
The config.js also relies on paths defined by ENV vars (or defaults) for this project.

We could generalize is even more and make it a generic runner, able to run arbitrary scripts and retrieve the output, but I'd say we'd probably be building a full blow CI system.

**Comment by vasco-santos** [comment]
I understand that generalize this solution more, at least at this moment seems not the best approach. 

I would like to have the `libp2p` benchmark tests in its own repo (`libp2p/benchmarks`). Moreover, I think it would be ok if we use this repo and infrastructure for both, and it gets the `libp2p/benchmark` tests from there. What do you think?

**Comment by litzenberger** [comment]
@Elexy I think the js-ipfs tests in the repo are fine, he just wants to pull another repo that will need to be synced by the runner.  @vasco-santos did I read that right?

**Comment by vasco-santos** [comment]
Yes, we should be able to get tests from other repos, such as `libp2p/benchmarks` repo, and in the future probably `ipld/benchmarks` as well.

**Comment by Elexy** [comment]
Currently we are `rsyncing` the tests over to the minion from the controller from this repo.In the new situation I'd need to run a `git clone` of a _specified repo_ before we can run the tests run the tests.

**Comment by Elexy** [comment]
As discussed in the latest [sync call](https://github.com/ipfs/benchmarks/issues/245) I want to present a proposal to allow the runner to run benchmark tests from a different repository.
The proposal is [here](https://docs.google.com/document/d/1oVgBkPc4iUg4-iKDiKBdWmcZc5Fqn_3MXp_nWvUHPeI/edit?usp=sharing).

**Comment by vasco-santos** [comment]
Thanks for the proposal @Elexy !
